lib/sycamore/sycamore/connectors/elasticsearch/elasticsearch_writer.py

```diff
@@ -0,0 +1,140 @@
+from typing import Any, Optional
+from sycamore.connectors.common import drop_types, flatten_data
+from typing_extensions import TypeGuard
+
+from sycamore.data.document import Document
+from sycamore.connectors.base import BaseDBWriter
+
+from elasticsearch import Elasticsearch, ApiError
+from elasticsearch.helpers import parallel_bulk
+
+
+@dataclass
+class ElasticClientParams(BaseDBWriter.ClientParams):
+    url: str
+    es_client_args: dict = field(default_factory=lambda: {})
+
+
+@dataclass
+class ElasticTargetParams(BaseDBWriter.TargetParams):
+    index_name: str
+    mappings: dict[str, Any] = field(
+        default_factory=lambda: {
+            "properties": {
+                "embeddings": {
+                    "type": "dense_vector",
+                    "dims": 384,
+                    "index": True,
+                    "similarity": "cosine",
+                },
+                "properties": {"type": "object"},
+            }
+        }
+    )
+    flatten_properties: bool = False
+
+    def compatible_with(self, other: BaseDBWriter.TargetParams) -> bool:
+        if not isinstance(other, ElasticTargetParams):
+            return False
+        if self.index_name != other.index_name:
+            return False
+        if self.mappings != other.mappings:
+            return False
+        return True
+
+
+class ElasticClient(BaseDBWriter.Client):
+    def __init__(self, client: Elasticsearch):
+        self._client = client
+
+    @classmethod
+    def from_client_params(cls, params: BaseDBWriter.ClientParams) -> "ElasticClient":
+        assert isinstance(params, ElasticClientParams)
+        client = Elasticsearch(params.url, **params.es_client_args)
+        return ElasticClient(client)
+
+    def write_many_records(self, records: list[BaseDBWriter.Record], target_params: BaseDBWriter.TargetParams):
+        assert isinstance(target_params, ElasticTargetParams)
+        assert _narrow_list_of_doc_records(records), f"Found a bad record in {records}"
+        with self._client:
+
+            def insert_into_index():
+                for r in records:
+                    yield {
+                        "_index": target_params.index_name,
+                        "_id": r.doc_id,
+                        "properties": r.properties,
+                        "embeddings": r.embeddings,
+                    }
+
+            for success, info in parallel_bulk(self._client, insert_into_index()):  # generator must be consumed
+                if not success:
+                    print(f"Insert Operation Unsuccessful: {info}")
+
+    def create_target_idempotent(self, target_params: BaseDBWriter.TargetParams):
+        assert isinstance(target_params, ElasticTargetParams)
+        try:
+            self._client.indices.create(index=target_params.index_name, mappings=target_params.mappings, timeout="30s")
+        except ApiError as e:
+            if e.status_code == 400:
+                return
```

Comment:
might wanna do a validation that the cause of the 400 was that the index exists already?